### DISCUSSION TITLE
Python vs C

### DISCUSSION THREAD

**Discussion by rhc54** [spack/spack#34071] [Q&A] [answered]
Hi folks

I am starting on a new tool (a dynamic scheduler) that my sponsor would someday like to see included in your distribution. I could write it in Python or in C - either way is fine with me. I already have two C-based packages in your distribution, so I'm quite familiar with how that works.

However, I have no experience with contributing a Python-based package. Indeed, one of my C-based packages (PMIx) has Python bindings available in it and we have been unable to get those included in the distribution (perhaps just lack of interest or effort).

I'm therefore wondering: do you have any advice as to choice of language here? Just glancing at things, it appears you have some Python-based packages - is there some guidance on how to do this, and/or an example you can point me to?

Any advice would be much appreciated!
Ralph

**Comment by alalazo** [comment, marked as answer]
Hi @rhc54 Both languages are well supported by Spack, so I'd say the choice is up to you. We interface pretty well with PyPI and we can download sources from there. Or we can download from GitHub directly.  You can look at literally any package starting with `py-*` to have some examples :slightly_smiling_face: Widely used (and moderately complex) Python packages are, for instance, `py-numpy`  or `py-scipy`. 

Also, I am curious why you have been unable to get the Python bindings for PMIx in. Let me know if you want to look again into that, I'd be glad to help.

**Reply by rhc54** [reply]
Thanks @alalazo - that is pretty much what I expected, but I appreciate the confirmation! I would welcome the assistance on the PMIx Python bindings as there has been growing interest in using them. I'll shift that discussion to an appropriate place.

One last question, if I may (apologies for this being off-topic for Spack): can you recommend a plugin manager for Python? I need something that can search/load plugins at startup. I have used `yapsy` in the past, but it hasn't been supported in some time and it has been years since I looked. Any advice would be welcome.

**Reply by alalazo** [reply]
That's the one I know of: https://pluggy.readthedocs.io/en/latest/ It's the plugin system used by `pytest`, so it should be fairly battle-tested.

**Reply by rhc54** [reply]
Thanks! Looks like that fits the bill nicely. I think I'll go forward with Python, then. Much appreciate the assistance!
